dklearn/wrappers.py

```diff
@@ -0,0 +1,256 @@
+
+    def __init__(self, name, dasks, base):
+        object.__setattr__(self, 'key', name)
+        object.__setattr__(self, 'dask', dasks)
+        object.__setattr__(self, '_base', base)
+
+    def _keys(self):
+        return [self.key]
+
+    def __repr__(self):
+        return 'DelayedEstimator<{0}>'.format(self.key)
+
+    def __hash__(self):
+        return hash(self.key)
+
+    def __dir__(self):
+        o = set(dir(type(self)))
+        o.update(self._base._get_param_names())
+        return list(o)
+
+    def __setattr__(self, attr, val):
+        raise TypeError("DelayedEstimator objects are immutable")
+
+    def __getattr__(self, attr):
+        if not attr.startswith('_') and attr in self._base._get_param_names():
+            return getattr(self._base, attr)
+        raise AttributeError("Attribute '{0}' not found".format(attr))
+
+
+def predict_chunk(est, X):
+    return est.predict(X)
+
+
+def decision_function_chunk(est, X):
+    return est.decision_function(X)
+
+
+def _maybe_concat(x):
+    return np.concatenate(x) if isinstance(x, list) else x
+
+
+@delayed(pure=True)
+def score_chunk(est, X, y, sample_weight=None):
+    X = _maybe_concat(X)
+    y = _maybe_concat(y)
+    sample_weight = _maybe_concat(sample_weight)
+    return est.score(X, y, sample_weight=sample_weight)
+
+
+class DaskWrapper(LazyDaskEstimator):
+    """Base class for classes that wrap a single estimator"""
+    def __init__(self, estimator):
+        self.estimator = estimator
+
+    def _to_delayed(self, x):
+        return DelayedEstimator(x.key, x.dask, clone(self))
+
+    def _update_estimator(self, x):
+        copy_to(x, self.estimator)
+        return self
+
+    def __dir__(self):
+        o = set(dir(type(self)))
+        o.update(vars(self))
+        for m in ['fit', 'predict']:
+            if not hasattr(self, m):
+                o.remove(m)
+        return list(o)
+
+    @if_delegate_has_method('estimator')
+    def fit(self, X, y, **kwargs):
+        compute = kwargs.pop("compute", True)
+        res = self._fit(X, y, **kwargs)
+        if compute:
+            return self._update_estimator(res.compute())
+        return self._to_delayed(res)
+
+    @if_delegate_has_method('estimator')
+    def predict(self, X, compute=True):
+        func = partial(predict_chunk, self.estimator)
+        if isinstance(X, da.Array):
+            assert X.ndim == 2
+            if len(X.chunks[1]) != 1:
+                X = X.rechunk((X.chunks[0], X.shape[1]))
+            res = X.map_blocks(func, chunks=(X.chunks[0],), drop_axis=1)
+        elif isinstance(X, db.Bag):
+            res = X.map_partitions(func)
+        elif is_list_of(X, Delayed):
+            func = delayed(func, pure=True)
+            res = [func(i) for i in X]
+            if compute:
+                return list(dask.compute(*res))
+            return res
+        else:
+            return func(X) if compute else delayed(func, pure=True)(X)
```

Comment:
My thought is wherever possible type in => type out. Lists of delayed and bags lose metadata, but are more general - arrays are best.